src/main/c/VipsImage.c

```diff
@@ -430,3 +430,47 @@ Java_com_criteo_vips_VipsImageImpl_release(JNIEnv *env, jobject obj)
         (*env)->SetLongField(env, obj, buffer_fid, (jlong) buffer);
     }
 }
+
+JNIEXPORT jint JNICALL Java_com_criteo_vips_VipsImageImpl_imageGetFormatNative(JNIEnv *env, jobject obj)
+{
+    VipsImage *im = (VipsImage *) (*env)->GetLongField(env, obj, handle_fid);
+    return vips_image_get_format(im);
+}
+
+JNIEXPORT void JNICALL Java_com_criteo_vips_VipsImageImpl_castUcharNative(JNIEnv *env, jobject obj, jboolean shift)
+{
+    VipsImage *im = (VipsImage *) (*env)->GetLongField(env, obj, handle_fid);
+    VipsImage *out = NULL;
+
+    if (vips_cast_uchar(im, &out, "shift", shift, NULL))
+    {
+        throwVipsException(env, "Unable to cast image to uchar");
+        return;
+    }
+    if (vips_image_get_format(out) != VIPS_FORMAT_UCHAR)
+    {
+        throwVipsException(env, "Bad format after casting to uchar");
+        return;
+    }
+    (*env)->SetLongField(env, obj, handle_fid, (jlong) out);
+    g_object_unref(im);
+}
```

Comment:
Maybe you can call Java_com_criteo_vips_VipsImageImpl_castNative() instead of duplicating code

Reply:
I'd rather keep the call to vips_cast_uchar because this is the call that I need and I'd rather not use VipsBandFormat in case it is wrong.